### ISSUE TITLE
Alignment broken.

### ISSUE THREAD

**Issue by nodeSpace** [emilk/egui#179]
Not sure if I'm doing something wrong but both `ui.vertical_centered()` and 
```rust
ui.allocate_ui(
    egui::vec2(
        ui.available_size_before_wrap_finite().x,
        ui.available_size_before_wrap_finite().y,
    ),|ui| ui.with_layout(
        egui::Layout::top_down(egui::Align::Center), |ui| {

        ui.text_edit_singleline(&mut root_node.content);
        ui.text_edit_singleline(test);
        ui.add(widgets::TextEdit::multiline(&mut root_node.content));
        //ui.text_edit_singleline(&mut root_node.content);
}));
```
produce this:
![image](https://user-images.githubusercontent.com/28955991/107905018-799c0400-6f1b-11eb-930e-140a25c501cd.png)

I suspect `ui.available_size_before_wrap_finite()` might be broken
This is with eframe, on windows, using desktop.

Edit: Same issue occurs on web as well. (egui 0.9.0)

**Comment by emilk** [comment]
I think this could be a bug in the top-down centered layout. Thanks for the report and the reproduce!

By the way: `egui::vec2(
        ui.available_size_before_wrap_finite().x,
        ui.available_size_before_wrap_finite().y,
    )` can be replaced by just `ui.available_size_before_wrap_finite()`.

**Comment by nodeSpace** [comment]
I think this is a bug in text_edit.rs, as other widgets seem to work fine. I'm having a hard time narrowing down where the issue is though. I *think* I narrowed it down to around this line:
https://github.com/emilk/egui/blob/fb2db4940e9181b46db2ead9089d7eafb92fbb91/egui/src/widgets/text_edit.rs#L243
but I don't really understand what's going on behind the scenes there. Could also be this section as well?
```rust
let (auto_id, rect) = ui.allocate_space(desired_size);
```

I took a screenshot with some debug_paint_cursor(), uploading it here incase it helps:
![image](https://user-images.githubusercontent.com/28955991/109469568-332ac700-7a3c-11eb-8944-600295ec0deb.png)
